### ISSUE TITLE
Get-CertificationAuthorityAcl is no longer returning ACLs on PowerShell 7.4.1

### ISSUE THREAD

**Issue by marcgoff** [Crypt32/PSPKI#208]
It seems like there is some change in Powershell 7.4+ that causes the Get-CertificationAuthorityAcl to no longer return data.

Expected behavior (using PSPKI 4.2.0 running on Powershell 7.2.18 on Windows Server 2016 Standard):
```
(Get-CertificationAuthority CA01.uad-dev.xxx.xxx |  Get-CertificationAuthorityAcl).Access

Rights            : Enroll
AccessControlType : Allow
IdentityReference : NT AUTHORITY\Authenticated Users
IsInherited       : False
InheritanceFlags  : None
PropagationFlags  : None

Rights            : ManageCA, ManageCertificates
AccessControlType : Allow
IdentityReference : BUILTIN\Administrators
IsInherited       : False
InheritanceFlags  : None
PropagationFlags  : None

Rights            : ManageCA, ManageCertificates
AccessControlType : Allow
IdentityReference : UAD-DEV\Domain Admins
IsInherited       : False
InheritanceFlags  : None
PropagationFlags  : None

Rights            : ManageCA, ManageCertificates
AccessControlType : Allow
IdentityReference : UAD-DEV\Enterprise Admins
IsInherited       : False
InheritanceFlags  : None
PropagationFlags  : None

```

Behavior using PSPKI 4.2.0 on Powershell 7.4.1 on Windows Server 2016 Standard:

```
(Get-CertificationAuthority CA01.uad-dev.xxx.xxx |  Get-CertificationAuthorityAcl -verbose )

Path Owner Access
---- ----- ------

```
The Get-CertificationAuthority command works fine, it could be something with the GetSecurityDescriptor method.

**Comment by Crypt32** [comment]
This module was never tested against PowerShell Core versions. The highest tested version is Windows PowerShell 5.1.

**Comment by Crypt32** [comment]
I've checked this issue and it appears that it is PowerShell 7.4 issue. The returned ACL is the same on both versions (Windows PowerShell 5.1 and PowerShell 7.4), however Windows PowerShell (and, PS Core prior to 7.4) does extra work to simplify the access to ACL:

```
PS C:\> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      5.1.26100.2161
PSEdition                      Desktop
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
BuildVersion                   10.0.26100.2161
CLRVersion                     4.0.30319.42000
WSManStackVersion              3.0
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1


PS C:\> $CA | Get-CAACL | gm


   TypeName: SysadminsLV.PKI.Security.AccessControl.CertSrvSecurityDescriptor

Name                            MemberType     Definition
----                            ----------     ----------
Access                          CodeProperty   System.Security.AccessControl.AuthorizationRuleCollection Access{get=...
Group                           CodeProperty   System.String Group{get=GetGroup;}
Owner                           CodeProperty   System.String Owner{get=GetOwner;}
Path                            CodeProperty   System.String Path{get=GetPath;}
Sddl                            CodeProperty   System.String Sddl{get=GetSddl;}
AccessRuleFactory               Method         System.Security.AccessControl.AccessRule AccessRuleFactory(System.Sec...
AddAccessRule                   Method         bool AddAccessRule(SysadminsLV.PKI.Security.AccessControl.CertSrvAcce...
AuditRuleFactory                Method         System.Security.AccessControl.AuditRule AuditRuleFactory(System.Secur...
Equals                          Method         bool Equals(System.Object obj)
```
Look at first 5 properties. These are compiled synthetic properties of type of  `CodeProperty` and are generated by PowerShell engine on a fly.

PowerShell 7.4 doesn't add them:
```
PS C:\> $PSVersionTable

Name                           Value
----                           -----
PSVersion                      7.4.6
PSEdition                      Core
GitCommitId                    7.4.6
OS                             Microsoft Windows 10.0.26100
Platform                       Win32NT
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0

PS C:\> $ca | Get-CAACL | gm

   TypeName: SysadminsLV.PKI.Security.AccessControl.CertSrvSecurityDescriptor

Name                            MemberType Definition
----                            ---------- ----------
AccessRuleFactory               Method     System.Security.AccessControl.AccessRule AccessRuleFactory(System.Security.…
AddAccessRule                   Method     bool AddAccessRule(SysadminsLV.PKI.Security.AccessControl.CertSrvAccessRule…
AuditRuleFactory                Method     System.Security.AccessControl.AuditRule AuditRuleFactory(System.Security.Pr…
Equals                          Method     bool Equals(System.Object obj)
```
There are no synthetic properties. But you still can access the ACL:

```
PS C:\> $acl = $ca | Get-CAACL
PS C:\> $acl.GetAccessRules($true,$true,[System.Security.Principal.NTAccount])

Rights            : Enroll
AccessControlType : Allow
IdentityReference : NT AUTHORITY\Authenticated Users
IsInherited       : False
InheritanceFlags  : None
PropagationFlags  : None

<...>
```
My bet it is a bug in PowerShell 7.4, because they indeed add synthetic properties to `FileSystemSecurity` for example. Looking at .NET inheritance tree, I see:
```
Object
└───ObjectSecurity
    └───CommonObjectSecurity
        ├───NativeObjectSecurity
        │   └───FileSystemSecurity
        └───CertSrvSecurityDescriptor
```
So, previously, they added synthetic properties to all instances of `CommonObjectSecurity` or even to `ObjectSecurity` instances, thus my [`CertSrvSecurityDescriptor`](https://www.pkisolutions.com/apidocs/pki/html/T_SysadminsLV_PKI_Security_AccessControl_CertSrvSecurityDescriptor.htm) which inherits from `CommonObjectSecurity` automatically received them out of the box. Somewhere in 7.4 they moved the logic to add these properties only to types that inherit from `NativeObjectSecurity` or even to a fixed list of types. So it really sounds like a PowerShell Core bug.

However, this is only visual representation issue on a console. By calling `GetOwner`, `GetAccessRules` methods on security descriptor object will let you access these properties in similar way as Windows PowerShell does. But if you do things like `Get-CA <...> | Get-CAACL | Add-CAACL | Set-CAACL` will work despite this visual issue.

Bottom line: there is very little I can do with this since these synthetic properties are added to the object after it is returned from PSPKI pipeline. You may want to ask this
